plugins/commhistory/src/commhistoryplugin.cpp

```diff
@@ -0,0 +1,218 @@
+        event.setIsEmergencyCall(handler.isEmergency());
+        event.setDirection(handler.isIncoming() ? CommHistory::Event::Inbound : CommHistory::Event::Outbound);
+
+        if (m_calls.isEmpty()) {
+            m_timer.start();
+        }
+        storeCall(&event);
+
+        m_calls.insert(handler.handlerId(), event);
+    }
+
+    void onVoiceCallStatusChanged(const AbstractVoiceCallHandler &handler)
+    {
+        CommHistory::Event &event = m_calls[handler.handlerId()];
+
+        switch (handler.status()) {
+        case AbstractVoiceCallHandler::STATUS_ACTIVE:
+            event.setStartTime(QDateTime::currentDateTime());
+            event.setEndTime(event.startTime());
+            event.setExtraProperty(activeCallProp, true);
+            storeCall(&event);
+            break;
+        case AbstractVoiceCallHandler::STATUS_DISCONNECTED:
+            event.setEndTime(event.startTime().addSecs(handler.duration()));
+            // No need to save the changes now, the voiceCallEnded will do it.
+            break;
+        default:
+            break;
+        }
+    }
+
+    void onVoiceCallEmergencyChanged(const AbstractVoiceCallHandler &handler)
+    {
+        CommHistory::Event &event = m_calls[handler.handlerId()];
+        event.setIsEmergencyCall(handler.isEmergency());
+        storeCall(&event);
+    }
+
+    void onVoiceCallDurationChanged(const AbstractVoiceCallHandler &handler)
+    {
+        CommHistory::Event &event = m_calls[handler.handlerId()];
+        event.setEndTime(event.startTime().addSecs(handler.duration()));
+    }
+
+    void storeCall(CommHistory::Event *event)
+    {
+        if (!event->isValid() && !m_eventModel.addEvent(*event)) {
+            qCWarning(voicecall) << "cannot add event to call history";
+        } else if (event->isValid() && !m_eventModel.modifyEvent(*event)) {
+            qCWarning(voicecall) << "cannot modify event in call history";
+        }
+    }
+
+    void storeCalls()
+    {
+        for (const QString &id : m_calls.keys()) {
+            CommHistory::Event &event = m_calls[id];
+            storeCall(&event);
+        }
+    }
+
+    void voiceCallEnded(const QString &id)
+    {
+        CommHistory::Event event = m_calls.take(id);
+        event.setIsMissedCall(event.direction() == CommHistory::Event::Inbound
+                              && !event.extraProperty(activeCallProp).isValid());
+        event.removeExtraProperty(activeCallProp);
+        if (event.isMissedCall()) {
+            event.setStartTime(QDateTime::currentDateTime());
+            event.setEndTime(event.startTime());
+        }
+        storeCall(&event);
+        if (m_calls.isEmpty()) {
+            m_timer.stop();
+        }
+    }
+};
+
+CommHistoryPlugin::CommHistoryPlugin(QObject *parent)
+    : AbstractVoiceCallManagerPlugin(parent), d(new Private)
+{
+    connect(&d->m_timer, &QTimer::timeout,
+            [this] () {
+                d->storeCalls();
```

Comment:
Optionally could connect directly to a method since this is doing just one method call.

Reply:
Well, `Private` class here is not inheriting QObject. I think to use a pointer to a member method, it must inherit from QObject…

That being said, I noticed after implementing all this, that having a Private class here is a bit over-engineered since it's a plugin. There is no requirement for ABI stability. If you think it's valuable, I can move all the Private class members to the CommHistoryPlugin one and add private methods there instead of hiding them behind the Private class.

Reply:
I'm fine either way. On extra alternative is just moving the storeCalls as private method on the base class.

Reply:
I'm keeping the Private class, but I've moved the `storeCalls()` implementation into the connection as it was called only from there.